### PR TITLE
[Prototype] Support composite ObjectSerializers, add XmlSerializer

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Prototype adding support for composite ObjectSerializers which will allow for `RestProxy`, and other code locations, to be created with multiple serialization formats for cases where a service uses multiple serialization formats. Additionally, `XmlSerializer` is added, the XML equivalent for `JsonSerializer` for JSON.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/serializer/CompositeSerializer.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * An internal type that comprises multiple {@link ObjectSerializer}s and adds functionality to determine which
+ * {@link SerializationFormat} to use each serialization and deserialize operation.
```
Comment (JonathanGiles) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/serializer/CompositeSerializer.java:19]:
```suggestion
 * {@link SerializationFormat} to use for each serialization and deserialize operation.
```

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java [hunk 2]
```diff
@@ -11,7 +11,13 @@
 /**
  * Generic interface covering basic JSON serialization and deserialization methods.
  */
-public interface JsonSerializer extends ObjectSerializer {
+public abstract class JsonSerializer extends ObjectSerializer {
```
Comment (JonathanGiles) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java:14]: If you're only changing this type from an interface to an abstract class for the sake of having an implementation of `supportsFormat`, have you considered keeping this as an interface and just having a default implementation?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/ObjectSerializer.java [hunk 3]
```diff
@@ -11,7 +11,13 @@
 /**
  * Generic interface covering serializing and deserialization objects.
  */
-public interface ObjectSerializer {
+public abstract class ObjectSerializer {
```
Comment (JonathanGiles) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/ObjectSerializer.java:14]: Why the change to abstract class here

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java:14]: I did, but I didn't want to give the ability to override this method as this serializer type is explicitly for JSON. If there are concerns with this change with regards to migration I'm willing to change this back if that simplifies things.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/ObjectSerializer.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/ObjectSerializer.java:14]: This can be changed back, was playing around with a few things and at one point abstract was what I needed.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/RestProxy.java [hunk 4]
```diff
@@ -68,24 +70,30 @@ public Object invoke(Object proxy, final Method method, Object[] args) {
...
+     * @param httpPipeline the HttpPipelinePolicy and HttpClient pipeline that will be used to send Http requests
+     * @param serializers the serializers that will be used to convert POJOs to and from request and response bodies
      * @param <A> the type of the Swagger interface.
-     *
      * @return a proxy implementation of the provided Swagger interface
+     * @throws IllegalArgumentException If {@code serializers} is null or empty.
      */
     @SuppressWarnings("unchecked")
-    public static <A> A create(Class<A> swaggerInterface, HttpPipeline httpPipeline, ObjectSerializer serializer) {
+    public static <A> A create(Class<A> swaggerInterface, HttpPipeline httpPipeline, ObjectSerializer... serializers) {
```
Comment (samvaity) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/http/RestProxy.java:94]: Would it help if we use the CompositeSerializer here directly ?
Specifically, for storage use cases where we want to support both JSON and XML formats.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/rest/RestProxyImpl.java [hunk 5]
```diff
@@ -243,7 +250,97 @@ public void updateRequest(RequestDataConfiguration requestDataConfiguration, Obj
...
+        if (lastIndex == -1) {
+            return SerializationFormat.JSON;
+        }
+
+        // Only XML and JSON are supported suffixes, there is no suffix for TEXT.
+        final String mimeTypeSuffix = subtype.substring(lastIndex + 1);
+        if ("xml".equalsIgnoreCase(mimeTypeSuffix)) {
+            return SerializationFormat.XML;
+        } else if ("json".equalsIgnoreCase(mimeTypeSuffix)) {
+            return SerializationFormat.JSON;
```
Comment (samvaity) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/rest/RestProxyImpl.java:302]: nit: Can we move this to a Util to make the class a bit leaner?

File: sdk/clientcore/core/src/test/java/io/clientcore/core/implementation/http/rest/SimpleXmlSerializable.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import io.clientcore.core.serialization.xml.XmlToken;
+import io.clientcore.core.serialization.xml.XmlWriter;
+
+import javax.xml.stream.XMLStreamException;
+import java.util.Objects;
+
+/**
+ * Test class implementing {@link XmlSerializable}.
+ */
+public final class SimpleXmlSerializable implements XmlSerializable<SimpleXmlSerializable> {
```
Comment (samvaity) [inline sdk/clientcore/core/src/test/java/io/clientcore/core/implementation/http/rest/SimpleXmlSerializable.java:18]: Simple or DefaultXmlSerializable provided by clientcore?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java [hunk 7]
```diff
@@ -3,26 +3,65 @@
...
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Type;
 
 /**
- * Generic interface covering basic JSON serialization and deserialization methods.
+ * Class providing basic JSON serialization and deserialization methods.
```
Comment (samvaity) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java:21]: Thoughts on keeping this interface and moving the implementation to DefaultJsonSerializer. 
edit: Offline chat, We may just want to expose ObjectSerializer for the customers to use their own implementations that can include both xml and Json formats.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/XmlSerializer.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * Class providing basic XML serialization and deserialization methods.
+ * <p>
+ * The implementation of this class is based on the usage of {@link XmlReader} and {@link XmlWriter}.
+ * <p>
+ * The deserialization methods only work with models implementing {@link XmlSerializable}. Or, in code terms, types that
+ * provide a static factory method {@code fromXml(XmlReader)}.
+ * <p>
+ * The serialization methods only work with complex types that implement {@link XmlSerializable}.
+ */
+public class XmlSerializer implements ObjectSerializer {
```
Comment (samvaity) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/XmlSerializer.java:30]: Does this function as default implementation for XML Serialization. How different is this from SimpleXMLSerializable, can the both be merged to be DefaulXMLSerializable?

File: sdk/clientcore/core/src/test/java/io/clientcore/core/implementation/http/rest/SimpleXmlSerializable.java (hunk 6 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/test/java/io/clientcore/core/implementation/http/rest/SimpleXmlSerializable.java:18]: This is a test class where we implement the XML streaming APIs from `io.clientcore.core.serializer.xml` to ensure XML support works in `RestProxy`.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java (hunk 7 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/JsonSerializer.java]: Changed to only expose `ObjectSerializer` for the time being

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/RestProxy.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/RestProxy.java:94]: For now, I'd keep `CompositeSerializer` as implementation and handle creating it from the `ObjectSerializer`s passed here. Using `RestProxy.create(Class<A>, HttpPipeline)` will be created with JSON and XML support, which should be sufficient for Storage.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/rest/RestProxyImpl.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/rest/RestProxyImpl.java:302]: I'm planning another PR in this area and will clean it up then. The `RestProxy` stuff is a bit complex being copied from `azure-core` where it has a sync and async variant. `clientcore` doesn't need to have `RestProxyBase`, `RestProxyImpl`, and `RestProxyUtil`, it can all be merged into `RestProxyImpl`.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/XmlSerializer.java (hunk 8 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/util/serializer/XmlSerializer.java]: `SimpleXmlSerializable` is an implementation of `XmlSerializable`, so it's a model type. This is the XML serializer implementation of `ObjectSerializer`.

**Review by samvaity — APPROVED** [review]
